### PR TITLE
(On Hold) Introduce SwapChain, refactor, and cleanup Instrumentation

### PR DESCRIPTION
**Pull request by ScottHolden**

### CODE REVIEW

**Comment by ScottHolden** [comment]
Added SkiaImageSwapChain which uses an image cache to replicate prior behaviour.

File: src/Trains.NET.Engine/GamePlay/GameBoard.cs [hunk 1]
```diff
@@ -79,15 +79,17 @@ public void GameLoopStep()
...
 
-            _gameUpdateTime.Start();
-            try
-            {
-                DoGameLoopStep();
-            }
-            catch (Exception)
+            using (var _ = _gameUpdateTime.Measure())
             {
+                _gameUpdateTime.Start();
```
Comment (davidwengier) [inline src/Trains.NET.Engine/GamePlay/GameBoard.cs:84]: Why doesn't Measure() call Start()?

Comment (ScottHolden) [reply src/Trains.NET.Engine/GamePlay/GameBoard.cs:84]: Haha, it does, I just missed removing this!

**Comment by davidwengier** [comment]
I've been trying to get this into a nice API, and its a struggle, but also, I don't get why this has to be so different. Why doesn't the WPF app just provider its own implementation of IImageFactory, and a CanvasWrapper?

With IImageFactory, the game is saying "Give me a canvas to draw on", so the factory can create the writable bitmap and set up the canvas etc. You're already doing that in DrawNext anyway. The IImage that is returned from IImageCanvas.Render() would be a WriteableBitmapFrameBuffer. That IImage is cached, and later the game calls canvas.DrawImage on it. Well that canvas is a CanvasWrapper that the GameElement passes in anyway, so it can pass in its own Canvas wrapper that knows how to handle the IImage that it also created.

Make sense? Would that work?

Essentially ImageCache is the "swap chain", and everything else works as you have it now, just with less actions. If image cache needs to get smarter to cache multiple buffers, thats fine (though I don't know what benefit that actually provides :P)

The DI not doing things correctly is another issue, but that can be fixed. In an ideal world though, to me this means the game itself provides everything it needs, though possible slow at times, and the WPF front end overrides a couple of those to provide versions that are optimized for that platform.

**Pull request closed without merging**